### ISSUE TITLE
Consul 1.13 cant disable TLS for gRPC Server when AutoEncrypt is in use (backwards compatibility & envoy usage broken)

### ISSUE THREAD

**Issue by pruiz** [hashicorp/consul#14253]
#### Overview

After upgrading to consul 1.13 our mesh services broke, due to envoy not being able to fetch configuration using xDS from consul agent. It looks like when AutoEncrypt is enabled (auto_encrypt.tls = true), even if ports.https is not defined, GRPC will be using TLS. This is similar to #14211, but the workaround proposed wont work in this case, as auto-encrypt forces usage of TLS on gRPC server no matter what.

Having gRPC server listen only on TLS breaks consul connect (envoy) support, as consul is unable to connect using TLS, nor present the required certificates with the boostrap configuration  generated by consul. This fails when using 'consul connect envoy' directly, and also when using nomad, especially with bridge networking, as in such cases nomad uses a socket to present a unix socket (proxied by nomad to the gRPC TCP listener from consul).

From what I've seen this issue has it's origin in two complementary changes applied to v1.13:

* https://github.com/hashicorp/consul/pull/12504 -> This PR introduced independent (per-listener) TLS settings, so we could set each interface using it's own TLS-related settings.
* https://github.com/hashicorp/consul/pull/13733 -> This PR removed the dependency between ports.https & enabling of grpc server's TLS, however this PR (IMHO) missed the additional effect of AutoEncrypt, effectively negating a way to use gRPC listener w/o TLS when AutoEncrypt is in use.

Additionally, even setting addresses.grpc to 'unix:///some.sock', so it listens over a unix socket instead of TCP, does not disable TLS on gRPC listener. Not sure this might be considered a bug or a feature.. ;)

PD: This other issue is probably related too: https://github.com/hashicorp/consul/issues/13124

#### Reproduction Steps

Use the following config on consul agents:

```
{
  "auto_encrypt": {
     "tls": true
  },
  "server": false,
  "encrypt": ...,
  "ports": {
     "http": 8500,
     "grpc": 8502
  },
  "tls": ... can be omitted or have any values ...
}
```

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 2
	services = 3
build:
	prerelease =
	revision = c6d0f9ec
	version = 1.13.1
	version_metadata =
consul:
	acl = enabled
	known_servers = 3
	server = false
runtime:
	arch = amd64
	cpu_count = 32
	goroutines = 87
	max_procs = 32
	os = linux
	version = go1.18.1
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 99
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 5887
	members = 24
	query_queue = 0
	query_time = 1```

</details>

<details>
  <summary>Server info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 2
	checks = 5
	services = 5
build:
	prerelease =
	revision = c6d0f9ec
	version = 1.13.1
	version_metadata =
consul:
	acl = enabled
	bootstrap = false
	known_datacenters = 3
	leader = true
	leader_addr = 172.22.70.11:8300
	server = true
raft:
	applied_index = 8890316
	commit_index = 8890316
	fsm_pending = 0
	last_contact = 0
	last_log_index = 8890316
	last_log_term = 5389
	last_snapshot_index = 8889355
	last_snapshot_term = 5389
	latest_configuration = [{Suffrage:Voter ID:ee56b4d5-2870-79fc-7dc5-73a2a22d0aa3 Address:172.22.70.11:8300} {Suffrage:Voter ID:bd3a8daf-f984-2254-f2dd-b2b699cee116 Address:172.22.70.12:8300} {Suffrage:Voter ID:72af041a-33d2-bcf5-7651-2e5581a73bd1 Address:172.22.70.13:8300}]
	latest_configuration_index = 0
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 5389
runtime:
	arch = amd64
	cpu_count = 2
	goroutines = 469
	max_procs = 2
	os = linux
	version = go1.18.1
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 99
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 5887
	members = 24
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 11875
	members = 9
	query_queue = 0
	query_time = 1```

</details>

### Operating system and Environment details

Debian 10, on amd64

### Log Fragments

This are some logs from our consul agents rejecting envoy xDS connection requestes due to not being TLS, but clear text:

Aug 17 23:00:44 host consul[2465206]: ==> Starting Consul agent...
Aug 17 23:00:44 host consul[2465206]:            Version: '1.13.1'
Aug 17 23:00:44 host consul[2465206]:         Build Date: '2022-08-11 19:07:00 +0000 UTC'
Aug 17 23:00:44 host consul[2465206]:            Node ID: '74d70774-2098-d526-00bb-78115ff46e09'
Aug 17 23:00:44 host consul[2465206]:          Node name: 'host'
Aug 17 23:00:44 host consul[2465206]:         Datacenter: 'ixn' (Segment: '')
Aug 17 23:00:44 host consul[2465206]:             Server: false (Bootstrap: false)
Aug 17 23:00:44 host consul[2465206]:        Client Addr: [127.0.0.1] (HTTP: 8500, HTTPS: -1, gRPC: 8502, DNS: 8600)
Aug 17 23:00:44 host consul[2465206]:       Cluster Addr: 172.22.70.204 (LAN: 8301, WAN: 8302)
Aug 17 23:00:44 host consul[2465206]:            Encrypt: Gossip: true, TLS-Outgoing: true, TLS-Incoming: false, Auto-Encrypt-TLS: true
[...]
Aug 18 00:36:28 host nsenter[2465206]: 2022-08-18T00:36:28.564Z [WARN]  agent: [core]grpc: Server.Serve failed to complete security handshake from "127.0.0.1:36402": tls: first record does not look like a TLS handshake

**Comment by pruiz** [comment]
I am working here (https://github.com/pruiz/consul/tree/pruiz/grpcserver-autoencrypt-fix) on a possible fix by introducing a new parameter to auto_encrypt to exclude gRPC listener from 'auto encryption'.. not sure if this would be the best approach though.

**Comment by ahjohannessen** [comment]
@rboyer @boxofrad @mkeeler Are you guys aware of these breaking changes?

**Comment by mkeeler** [comment]
@ahjohannessen We became aware of this yesterday and are working on a fix. My current thinking is that in the short term (1.13.x - almost certainly 1.13.2) we will introduce a new configuration to prevent the gRPC listener from using the auto-encrypt certificate.

Longer term, it feels like users need more control over gRPC listeners. I am thinking that we could allow users to configure both unencrypted and encrypted gRPC listeners and that way there is no implicit enabling of TLS.

**Comment by jkirschner-hashicorp** [comment]
We're also working on updating our upgrade documentation to reflect our understanding of the problem and present recommendations.

**Comment by mkeeler** [comment]
@pruiz Your solution looks to be exactly what I was thinking we should do for the short term fix. Were you planning on opening a pull request for the changes?

**Comment by ahjohannessen** [comment]
@mkeeler Seems like a sound approach. Hopefully @pruiz can make a PR :)

**Comment by pruiz** [comment]
@mkeeler @ahjohannessen yeah I am working on it right now.. got an initial version working, but I need to make some cleanups.. hopefully I will be able to submit a PR tonite or maybe tomorrow.

**Comment by pruiz** [comment]
Hi,

I've just created a couple of PRs:
* https://github.com/hashicorp/consul/pull/14268 => fix against release/1.13.x branch
* https://github.com/hashicorp/consul/pull/14269 => fix against main branch

Regards

**Comment by jkirschner-hashicorp** [comment]
We just [updated the 1.13 upgrade documentation to call out this breaking change](https://www.consul.io/docs/upgrading/upgrade-specific#service-mesh-deployments-using-auto-encrypt-or-auto-config) and advise not upgrading to 1.13 _yet_ if using auto-encrypt or auto-config. I'll look into adding a notice to the changelog as well.

Thank you, @pruiz and @ahjohannessen, for flagging this and sharing agent configuration snippets that helped the Consul team investigate this in parallel with your own efforts. And thank you @pruiz for proposing and submitting a PR for us to consider.

**Comment by ahjohannessen** [comment]
@jkirschner-hashicorp Great. Looking forward to a patch release 👍

**Comment by mkeeler** [comment]
The fix for this issue has been merged and will be available in the 1.13.2 release when it is made. The fix ended up looking like:

If you want to use the auto-encrypt/auto-config certificate for gRPC TLS (current undesirable behavior)
```
tls {
   grpc {
      use_auto_cert = true
   }
}
```

If you don't want that behavior the defaults should work for you or you can be extra explicit and disable it in config:
```
tls {
   grpc {
      use_auto_cert = false
   }
}
```

**Comment by Xuxe** [comment]
Hey, 
i see similar behavior on Consul 1.13.2 with Nomad 1.4.0. In my scenario i have deployed 3 VMs running Nomad Server + Client and Consul Server + Agent per VM. I have extra VMs for Nomad Clients + Consul Agents.  I use auto_encrypt for Consul. 

On my consul server nodes i have:
```
auto_encrypt = {
  allow_tls = true
 }
```
On the consul client nodes i have:
```
auto_encrypt = {
  tls = true
}
```

TLS settings of consul are as followed:
```
  tls {
     grpc {
      use_auto_cert = false
      verify_incoming = false
    }
    
   https {
      verify_incoming = false
    }
   
   defaults {
        verify_incoming = false
        verify_outgoing = true
        ca_file = "/opt/consul/consul-agent-ca.pem"
        key_file = "/opt/consul/consul-server-key.pem"
        cert_file = "/opt/consul/consul-server.pem"
    }

    internal_rpc {
      verify_server_hostname = true
      verify_incoming = true
    }
}
```
Consul Client Nodes don't have the key_file and cert_file set. 

When i deploy a work load in Nomad on the server nodes i get issues with the Envoy sidecar task. As soon as i move the workload with the sidecar to dedicated client nodes (where Consul is running in client mode) it works flawlessly? 

Envoy presents:
`[2022-10-17 17:22:45.866][1][warning][config] [./source/common/config/grpc_stream.h:196] DeltaAggregatedResources gRPC config stream closed since 636s ago: 14, upstream connect error or disconnect/reset before headers. reset reason: connection termination`
Consul logs:
`Oct 17 17:24:39 nomad-03 consul[3108781]: 2022-10-17T17:24:39.626Z [WARN]  agent: [core]grpc: Server.Serve failed to complete security handshake from "127.0.0.1:50986": tls: first record does not look like a TLS handshake`

Somehow it looks like even with the GRPC work a round applied that the settings are ignored on Consul server nodes? 

I am not sure if its really the same/related issue but looks very likely.
